cpp/src/gandiva/gdv_string_function_stubs.cc

```diff
@@ -264,6 +265,304 @@ const char* gdv_fn_lower_utf8(int64_t context, const char* data, int32_t data_le
+
+        // If it is an invalid utf8 character, UTF8Decode evaluates to false
+        if (!is_valid_utf8_char) {
+          gdv_fn_set_error_for_invalid_utf8(context, data[i]);
+          *out_len = 0;
+          return "";
+        }
+
+        char hex_string[20];
+
+        snprintf(hex_string, y, "%.4X", char_codepoint);
+
+        for (int32_t xx = 0; xx < 4; xx += 4) {
+          char pp = hex_string[xx];
+          hex_string[xx] = hex_string[xx + 2];
+          hex_string[xx + 2] = pp;
+
+          pp = hex_string[xx + 1];
+          hex_string[xx + 1] = hex_string[xx + 3];
+          hex_string[xx + 3] = pp;
+        }
+
+        memcpy(tmp, hex_string, 4);
+        tmp += 4;
+      } else if (char_len == 3) {
+        const auto* in_char = (const uint8_t*)(data + i);
+
+        // Decode the multibyte character
+        bool is_valid_utf8_char =
+            arrow::util::UTF8Decode((const uint8_t**)&in_char, &char_codepoint);
+
+        // If it is an invalid utf8 character, UTF8Decode evaluates to false
+        if (!is_valid_utf8_char) {
+          gdv_fn_set_error_for_invalid_utf8(context, data[i]);
+          *out_len = 0;
+          return "";
+        }
+
+        char hex_string[20];
+
+        snprintf(hex_string, z, "%.8X", char_codepoint);
+
+        for (int32_t xx = 0; xx < 8; xx += 4) {
+          char pp = hex_string[xx];
+          hex_string[xx] = hex_string[xx + 2];
+          hex_string[xx + 2] = pp;
+
+          pp = hex_string[xx + 1];
+          hex_string[xx + 1] = hex_string[xx + 3];
+          hex_string[xx + 3] = pp;
+        }
+
+        memcpy(tmp, hex_string, 8);
+        tmp += 8;
+      } else if (char_len == 4) {
+        const auto* in_char = (const uint8_t*)(data + i);
+
+        // Decode the multibyte character
+        bool is_valid_utf8_char =
+            arrow::util::UTF8Decode((const uint8_t**)&in_char, &char_codepoint);
+
+        // If it is an invalid utf8 character, UTF8Decode evaluates to false
+        if (!is_valid_utf8_char) {
+          gdv_fn_set_error_for_invalid_utf8(context, data[i]);
+          *out_len = 0;
+          return "";
+        }
+
+        char hex_string[20];
+
+        snprintf(hex_string, z, "%.8X", char_codepoint);
+
+        for (int32_t xx = 0; xx < 8; xx += 4) {
+          char pp = hex_string[xx];
+          hex_string[xx] = hex_string[xx + 2];
+          hex_string[xx + 2] = pp;
+
+          pp = hex_string[xx + 1];
+          hex_string[xx + 1] = hex_string[xx + 3];
+          hex_string[xx + 3] = pp;
+        }
+
+        memcpy(tmp, hex_string, 8);
+        tmp += 8;
+      }
+    }
+    *out_len = static_cast<int32_t>(tmp - out);
+    return out;
+  }
+  return "";
+}
+
+// converts utf16 to utf8 encoding
+GANDIVA_EXPORT
+const char* gdv_fn_decode(int64_t context, const char* data, int32_t data_len,
+                          const char* charset, int32_t charset_len, int32_t* out_len) {
```

Comment:
I don't see any utf8 encoding in this function. With non-ASCII tests you would probably notice the problem.

Reply:
I have probably added a unit test for non-ascii which returns null string.

Reply:
I'm talking about a test with valid non-ASCII data, which would return successfully.